File: docs/scripting/samples/csharp-scripts/examples/Remove_Steps.cs
/**
 * Remove Steps
 * 
 * Removes specific protocol steps from plates dynamically
 */


using System;
using System.Linq;
using HRB.Cellario.Scripting.API;
using System.Collections.Generic;

namespace Customer.Scripting
{
    public class RemoveSteps : AbstractScript
    {
        public override void Execute(IScriptingApi api)
        {
        	//only execute this script once -- when the first plate in the thread is introduced
        	if (api.CurrentPlate.PlateNumber != 1) return;
        	
            var plates = api.GetPlatesForCurrentThread();
            foreach (var plate in plates)
            {
                if (plates.Any(p => p.StartingLocation.ResourceName == plate.StartingLocation.ResourceName &&
                                    p.StartingLocation.Stack == plate.StartingLocation.Stack &&
                                    p.StartingLocation.Position == plate.StartingLocation.Position &&
                                    p.PlateNumber < plate.PlateNumber))
                {
                    RemoveStep(plate, "SPIN");
                    RemoveStep(plate, "PIERCE");
                }
                if (plates.Any(p => p.StartingLocation.ResourceName == plate.StartingLocation.ResourceName &&
                                    p.StartingLocation.Stack == plate.StartingLocation.Stack &&
                                    p.StartingLocation.Position == plate.StartingLocation.Position &&
                                    p.PlateNumber > plate.PlateNumber))
                    RemoveStep(plate, "SEAL");
            }
        }
        
        public void RemoveStep(IScriptingPlate plate, string stepName)
        {
            var ind = plate.RemainingSteps.ToList().FindIndex(x => x.StepName.ToUpper().Contains(stepName));
            var moveStep = plate.RemainingSteps.ElementAt(ind + 1);
            plate.RemainingSteps.Remove(moveStep);
            var step = plate.RemainingSteps.ElementAt(ind);
            plate.RemainingSteps.Remove(step);
        }
    }
}

File: docs/scripting/samples/python-scripts/python-examples/Remove_Steps.py
"""
Remove Steps

Removes specific protocol steps from plates dynamically
"""

#Cellario modules
from HRB.Cellario.Scripting import *
from HRB.Cellario.Scripting.API import *


def Execute(api : PythonScriptingApi):
    if (api.CurrentPlate.PlateNumber != 1):
        return

    plates = list(api.GetPlatesForCurrentThread())
    
    for plate in plates:
        remove_step(api, plate, "SPIN")


# Removes the first instance of the step from this plate.
def remove_step (api : PythonScriptingApi, plate, step_name):
    remaining_steps_list = list(plate.RemainingSteps)
    step = next(s for s in remaining_steps_list if s.StepName.upper() == step_name)
    ind = remaining_steps_list.index(step)
    plate.RemainingSteps.Remove(step)
    api.Messaging.WriteDiagnostic(ScriptLogLevel.Normal, "Removing step: " + step.StepName)
  